myhoard/restore_coordinator.py

```diff
@@ -1128,13 +1137,39 @@ def _rename_prefetched_binlogs_back(self, binlogs):
                 last_renamed_index = remote_index - 1
         self.update_state(last_flushed_index=last_renamed_index, last_renamed_index=last_renamed_index)
 
+    def _sql_thread_not_running(self, slave_status):
+        return slave_status["Slave_SQL_Running"] != "Yes"
+
+    def _start_sql_thread(self, cursor):
+        # Try to start SQL thread with an incremental delay
+        if self.sql_thread_restart_count > 0:
+            factor = 1.5 ** self.sql_thread_restart_count
+            iteration_delay = min(self.SQL_THREAD_RESTART_INIT_INTERVAL * factor, self.SQL_THREAD_RESTART_MAX_INTERVAL)
+            if self.sql_thread_restart_time is not None:
+                time_passed = time.monotonic() - self.sql_thread_restart_time
+                if time_passed < iteration_delay:
+                    return False
+        self.sql_thread_restart_time = time.monotonic()
+        self.log.warning("SQL thread is not running. Running 'START SLAVE SQL_THREAD'")
+        self.sql_thread_restart_count += 1
+        cursor.execute("START SLAVE SQL_THREAD")
+        return True
+
     def _check_sql_slave_status(self):
         expected_range = self.state["current_executed_gtid_target"]
         expected_index = self.state["current_relay_log_target"]
 
         with self._mysql_cursor() as cursor:
             cursor.execute("SHOW SLAVE STATUS")
             slave_status = cursor.fetchone()
+
+            # Sometimes mysqld can be OOM-killed during the backup restoration.
+            # In that case we should try restarting it.
+            if self._sql_thread_not_running(slave_status):
```

Comment:
I checked the test failure and based on that the issue here is likely that this start is performed before checking whether the target position has been reached or not. When restoration is done in PITR mode it only restores up until certain point in a binlog and then stops the sql thread but now this goes and starts it again and it applies more transactions that is expected. Fix should be to only start the thread before returning false later on in this function (to simplify the call maybe do `_start_sql_thread` => `_start_sql_thread_if_not_running` and since the call the fetch slave status again is unnecessary you can just add simple `self._start_sql_thread_if_not_running()` call to the two relevant places later on in this function.

Reply:
Updated as suggested, hopefully I understood the code correctly.